### ISSUE TITLE
define behaviors of the common ConstrainablePattern Interfaces 

### ISSUE THREAD

**Issue by ShijunS** [w3c/mediacapture-fromelement#48]
The spec currently doesn't cover behaviors for the ConstrainablePattern Interfaces.  As example, here are some questions with applyConstraints()
- Whether the "frameRate" constraint can be applied to video track captured from video element and canvas (might be relevant to issue #43)
- Whether apps want to capture a lower resolution video from an UHD/4k-video source, and whether the "width" and "height" constraints should be honored in that case
- Whether the "channelCount" constraint should be applied for captured audio tracks.  As an potential use case, we might need to convert multi-channel audio to stereo before we send it out with peerconnection.  
- How about other audio constraints: volume, sampleRate, and sampleSize - although it might require audio DSP operations so leading to more complexity in implementations.  

In case any of the common ConstrainablePattern Interfaces is irrelevant, the spec should call that out explicitly.

**Comment by martinthomson** [comment]
Yes, I think that this is important.  I think that the only things we want to honor here is the frameRate constraint.  The rest can remain unsupported.

**Comment by alvestrand** [comment]
If one desires to make a 640x480 recording of an UHD video element (<video> to MediaStreamRecorder), how would one specify that if "width" and "height" cannot be constrained?

(It's OK to say "can't do this", but I don't like the idea of making the spec "can't ask for this").

**Comment by Pehrsons** [comment]
In my opinion you cannot constrain the decoder because it will output UHD frames always - as they were encoded. It would become a processing step rather than a configuration (by constraining the source) step.

I think MediaRecorder should have an API for this instead.

**Comment by alvestrand** [comment]
In the scenario I sketched, there might not be a decoder (the UHD-sized video element might be showing animation from a Canvas). Or there might be one.
The video element what it produces into a MediaStreamTrack. A MediaStreamTrack has a constrainable interface.
Why would we specify two APIs for doing the same thing?

In all cases, the UHD frames will be fed to a rescaler somewhere in the system. The question is just about how we specify the API whereby the user can say "I want this to be rescaled".

**Comment by yell0wd0g** [comment]
To bring the Constrainable Pattern to the `CanvasCaptureMediaStreamTrack` and drawing from my experience with ImageCapture https://github.com/w3c/mediacapture-image/pull/146, I think roughly here we'd need to define a supported constraints dictionary as sth like:

```
partial dictionary MediaTrackSupportedConstraints {
 boolean width = false;
 boolean height = false;
 boolean aspectRatio = false;
 boolean frameRate = true;
 boolean facingMode = false;
 boolean volume = false;
 boolean sampleRate = false;
 boolean sampleSize = false;
 boolean echoCancellation = false;
 boolean latency = false;
 boolean channelCount = false;
 boolean deviceId = false;
 boolean groupId = false;
};
```

and also the behaviour of the rest of the Constrainable Pattern methods that are needed as well, i.e. these guys:

```
MediaTrackCapabilities getCapabilities ();
MediaTrackConstraints getConstraints ();
MediaTrackSettings getSettings ();
Promise<void> applyConstraints (optional MediaTrackConstraints constraints);
```

And to round the change, `captureStream()` should accept a [`MediaTrackConstraintSet`](https://www.w3.org/TR/mediacapture-streams/#dictionary-mediatrackconstraints-members) a la `getUserMedia()`.

We should also have a similar analysis for the "standard" MediaStreamTrack produced by `captureStream()` on an HTMLMediaElement.

ImageCapture has a rendering of my proposed PR in https://rawgit.com/Miguelao/mediacapture-image/pr001_constrainable_pattern/index.html

**Comment by martinthomson** [comment]
Maybe width and height are OK to change.  I'd have to double check with @jan-ivar though on the latest viewpoint on those regarding processing in the track (as opposed to having the source just change its capture size).

**Comment by martinthomson** [comment]
If we only have one constrainable value, `MediaTrackConstraintSet` is overkill, but if the intent is to make things consistent, then fine.  Every time that pattern spreads its shadow the web dies a little more, but it was already pestilent when we got here I guess.

**Comment by alvestrand** [comment]
as to the mediaTrackSupportedConstraints thing - I think you would want to return this in Capabilities (give no alternatives for things you can't alter) rather than supportedConstraints (which ar esupposed to list things where the name is understood by the platform).

given that CanvasCaptureMediaStreamTrack is a subclass of MediaStreamTrack, it already inherits the constrainable-pattern functions. You don't get to remove features when inheriting.

**Comment by jan-ivar** [comment]
The Constrainable pattern was designed for device discovery and sharing. It's complex, because A) people's devices may not suit your needs, and B) devices you get may already be in use, effectively shared with another party (another tab) - without any knowledge of other parties - which puts a cramp on available settings. It's an API designed around compromise and negotiation.

A rescaler is the opposite of that. If you ask for 800x600, you'll always get it. Every time. Exactly.

This makes the Constrainable pattern overkill and a terrible API for a rescaler.

It's also a weak API, because the pattern allows huge variance in browser implementations. Maybe you'll get `OverConstrainedError` if you used `exact`, maybe it'll succeed with different values than you asked for if you used `ideal`. Instead, you just want all browsers to rescale. Every time. Exactly.

> In case any of the common ConstrainablePattern Interfaces is irrelevant, the spec should call that out explicitly.

I'd say on the contrary, unless a source explicitly defines constrainable properties, then there are none implied by simply pulling in a pattern. E.g. None of the properties `width`, `height` or `frameRate` come with the ConstrainablePattern.

Also, IMHO we should start with needs, not go looking for functionality to implement out of some need for symmetry.

From the needs I've heard, something straightforward like what @pehrsons mentions sounds better to me.

**Comment by EliasHasle** [comment]
A use case of captureStream is to record video of rendered graphics on a canvas. Naturally, with graphics rendering the FPS may be variable. For realtime graphics a variable output FPS is acceptable, but for graphics rendered on given time steps, but with variable processing time, such as for a movie (could be using raytracing or other expensive techniques), it should be possible to render frames one by one and specify a fixed framerate. I actually expected this behavior for captureStream(0), but found no way to control the output framerate. Maybe requestFrame could take a dt or timestamp parameter, to override the realtime default? Or maybe there could be a way to apply constraints?

Syncing with an audio stream would then be another challenge.

I understand that realtime streams may be prioritized, but then maybe the output framerate control belongs in MediaRecorder.

**Comment by jan-ivar** [comment]
We've identified this as a web compat issue in [bug 1942763 (comment 5)](https://bugzilla.mozilla.org/show_bug.cgi?id=1942763#c5).

With https://jsfiddle.net/jib1/k1bLrsnx/16/ we see both Chrome and Safari reporting `width` and `height`, whereas Firefox doesn't. I propose we add those two to this spec.¹

---
<sub>1. Chrome additionally reports `frameRate`,  but its value seems unreliable and [unrelated to anything happening in the canvas](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/mediastream/media_stream_video_track.cc;l=1137;drc=1f9f3d7d4227fc2021915926b0e9f934cd610201), so I propose we leave it out.</sub>

**Comment by youennf** [comment]
Maybe media capture-main should state that all video tracks should by default have width, height and maybe frameRate, unless otherwise specified by more specific specs.

**Comment by jan-ivar** [comment]
How about adding to [§ 17.4 Defining a new source of MediaStreamTrack](https://w3c.github.io/mediacapture-main/getusermedia.html#defining-a-new-source-of-mediastreamtrack) ("At a minimum, a new source of [MediaStreamTrack](https://w3c.github.io/mediacapture-main/getusermedia.html#dom-mediastreamtrack) will need to:"):

- _"for [MediaStreamTrack](https://w3c.github.io/mediacapture-main/getusermedia.html#dom-mediastreamtrack)s of kind `"video"`, define the constrainable properties `width`, `height` and `aspectRatio`, and describe how they work as settings (they need not be constrainable),"_

It probably remains the responsibility of each source spec to define how their constrainable properties work. For instance, [width](https://w3c.github.io/mediacapture-screen-share/#dfn-width) in screen-capture is quite different from [width](https://w3c.github.io/mediacapture-main/getusermedia.html#def-constraint-width) for cameras.

Unsure about frameRate. What would it be for canvas?
